flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TableSourceUtils.java

```diff
@@ -0,0 +1,240 @@
+				&& proctimeAttribute.map(attr -> !attr.equals(col.getName())).orElse(true))
+			.collect(Collectors.toList());
+
+		Map<TableColumn, Integer> columnsToPhysicalIndices = TableSourceUtils.computePhysicalIndices(
+			columnsWithoutTimeAttributes.stream(),
+			tableSource.getProducedDataType(),
+			nameRemapping
+		);
+
+		return logicalColumns.stream().mapToInt(logicalColumn -> {
+			if (proctimeAttribute.map(attr -> attr.equals(logicalColumn.getName())).orElse(false)) {
+				verifyTimeAttributeType(logicalColumn, "Proctime");
+
+				if (streamMarkers) {
+					return TimeIndicatorTypeInfo.PROCTIME_STREAM_MARKER;
+				} else {
+					return TimeIndicatorTypeInfo.PROCTIME_BATCH_MARKER;
+				}
+			} else if (rowtimeAttributes.contains(logicalColumn.getName())) {
+				verifyTimeAttributeType(logicalColumn, "Rowtime");
+
+				if (streamMarkers) {
+					return TimeIndicatorTypeInfo.ROWTIME_STREAM_MARKER;
+				} else {
+					return TimeIndicatorTypeInfo.ROWTIME_BATCH_MARKER;
+				}
+			} else {
+				return columnsToPhysicalIndices.get(logicalColumn);
+			}
+		}).toArray();
+	}
+
+	private static void verifyTimeAttributeType(TableColumn logicalColumn, String rowtimeOrProctime) {
+		if (!hasFamily(logicalColumn.getType().getLogicalType(), LogicalTypeFamily.TIMESTAMP)) {
+			throw new ValidationException(String.format(
+				"%s field '%s' has invalid type %s. %s attributes must be of a Timestamp family.",
+				rowtimeOrProctime,
+				logicalColumn.getName(),
+				logicalColumn.getType(),
+				rowtimeOrProctime));
+		}
+	}
+
+	private static Map<TableColumn, Integer> computePhysicalIndices(
+			Stream<TableColumn> columns,
+			DataType physicalType,
+			Function<String, String> nameRemappingFunction) {
+		if (DataTypeUtils.isCompositeType(physicalType)) {
+			TableSchema physicalSchema = TypeConversions.expandCompositeTypeToSchema(physicalType);
+			return computeInCompositeType(columns, physicalSchema, nameRemappingFunction);
+		} else {
+			return computeInSimpleType(columns, physicalType);
+		}
+	}
+
+	private static Map<TableColumn, Integer> computeInCompositeType(
+			Stream<TableColumn> columns,
+			TableSchema physicalSchema,
+			Function<String, String> nameRemappingFunction) {
+		return columns.collect(
+			Collectors.toMap(
+				Function.identity(),
+				column -> {
+					String remappedName = nameRemappingFunction.apply(column.getName());
+
+					int idx = IntStream.range(0, physicalSchema.getFieldCount())
+						.filter(i -> physicalSchema.getFieldName(i).get().equals(remappedName))
+						.findFirst()
+						.orElseThrow(() -> new TableException(String.format(
+							"Could not map %s column to the underlying physical type %s. No such field.",
+							column.getName(),
+							physicalSchema
+						)));
+
+					LogicalType physicalFieldType = physicalSchema.getFieldDataType(idx).get().getLogicalType();
+					LogicalType logicalFieldType = column.getType().getLogicalType();
+
+					if (!physicalFieldType.equals(logicalFieldType)) {
```

Comment:
Hi @dawidwys , thanks for fixing this problem. 

I looked into the changed code. The test is failed because it uses the new `computePhysicalIndices` in table-common. In blink planner, we uses `PlannerTypeUtils#isAssignable` to assert whether the physical and logical type are equal. The `isAssignable` is very loose which only checks type root, but `LogicalType#equals` is the most strict which also checks field names, precisions and nullable. In previous, the old planner check physcial and logical type using TypeInformation which is the same to `isAssignable`.

From my point of view, we don't need to check precisions and nullable, because we are using TypeInformation in runtime, e.g. BigDecimal. It's fine it uses (38, 18) in reader, because we will convert it into the correct precision (precision from DDL) in `SourceConversion` via `DataFormatConverter`.

Reply:
Converting precisions sounds wrong to me. If applications work in financial data, there should be not conversion of the data at all. So produced and expected type must match in this case. Of course, we can relax this constraint for the legacy types case and also ignore fields names of a row type but ignoring precision and scale should not be skipped.

Reply:
Hi @twalthr , I'm fine with relax contraint only for legacy types. But this requires that we have to update all the connectors (including users') to the new type system at once. From my point of view, it's safe to ignore precision, because 
1) all the exteranl data format (e.g. BigDecimal, Timestamp) can retain the full precision informaiton for data. So converting it to the precision defined in DDL doesn't lose information. 
2) even if users give a precision via `getProducedDataType`, it is hard to guarantee the produced data is following the precision.

Reply:
First, this problem needs to be fixed, because many of the current connectors are based on TypeInformation.

I think the ideal fix would be implicit type transformation, will add a cast before sink, which is in our 1.11 goal.

I'm fine with relax constraint only for legacy types too.
@wuchong But I don't understand why it requires update all connectors?
We can just check if it is a legacy type, we can respect type from select table instead of this legacy type, and use in validator and execution.

Reply:
@JingsongLi , we can't distinguish whether the connector is using legacy type or new type for `Types.SQL_TIMESTAMP` which will be converted to `DataTypes.TIMESTAMP(3).bridgeTo(Timestamp.class)`.

Reply:
I discussed this again with @dawidwys. We would like to propose to check `LogicalTypeChecks#areTypesCompatible` or allow legacy decimal type; all of that logically.

We don't need to change all connectors at once, instead we force users to define the precision and scale they expect which is what the connector currently supports. For example: If the CSV source currently returns `Types.SQL_TIMESTAMP`, users can simply specify `TIMESTAMP(3)`. In the future, we when we upgrade our CSV source, users will still get precision 3. If we would not check the precision at all, we would break pipelines that defined `TIMESTAMP(9)` because suddenly more precision will be added by the runtime when we upgrade the CSV source.

Reply:
> If we would not check the precision at all, we would break pipelines that defined TIMESTAMP(9) because suddenly more precision will be added by the runtime when we upgrade the CSV source.

That's true in old planner. But blink planner uses the logical types defined in DDL, so it is forward compatible for blink planner. 

But I'm fine with the proposed approach, this is also the initial approach of FLINK-15124. If we go with this approach, I would suggest to throw a meaningful exception message to remind users to upgrade `TableSource#getProducedDataType` of the connector.


Btw, there is a corresponding problem with the sink. Shall we check precisions for sinks too? Currenlty, it's hard to upgrade to new type system for `UpsertStreamTableSink` and `RetractStreamTableSink`.

Reply:
My comment is also true for the Blink planner. If the DDL defines `TIMESTAMP(9)` (and the produced type is usually derived from the DDL with some modification) and the connector cannot produce what is requested, we should throw an exception.

Reply:
So what about the sinks? @twalthr

Reply:
Yes, we should also check the sinks. I hope the entire type business will find an end once we have the new source/sink interfaces. FLIP-27 seems to make good progress.

Reply:
@wuchong In my opinion it would definitely make sense for sinks as well.

Reply:
In my opinion, the core issue is:
```
CREATE TABLE my_kafka (
  f0 STRING,
  f1 TIMESTAMP(3),
  f2 DECIMAL -- or DECIMAL(38, 18)?
) WITH (
  connector.type = 'kafka'
)
```
```
INSERT INTO my_kafka SELECT 'my_key_0', CURRENT_TIMESTAMP, 5.45;
```

Note, in blink planner, `'my_key_0'` will produce `CHAR(8)`, `CURRENT_TIMESTAMP` will produce `TIMESTAMP(6)`, `5.45` will produce `DECIMAL(3, 2)`.
It means that we need to write data with different precision. Even different types(char and varchar).
Note, this is a behavior that will have been supported by legacy planner.

I think we have three choice:
1. Strict check, throw exception. Tell user to cast by themselves.
2. Relax check, because our current format is compatible with precision, which is also our current implementation.
3. Add an implicit cast.

Reply:
First of all, what you are describing @JingsongLi affects sinks. Moreover it affects the mapping of logical types of a query to logical types of a sink, not physical-logical mapping of sink's types. 

The resulting `my_kafka` table would have:
* TableSchema('f0' STRING, f1 TIMESTAMP(3), f2 DECIMAL(38,18))
* consumedDataType( ??? )  <- could have something different from the schema, e.g. different precisions

Does that make sense?

What we discussed is a check that the description of a TableSink is consistent, a.k.a logical-physical mapping (1). Not a check if a query can be written into a Sink (2).

For (1) as I said before I think we should be strict (minus Row names and legacy decimal type)
For (2) I think this should follow any other logical operation. It should try to add **implicit** casts (if possible). If implicit casts are not possible, we throw an exception. Users still have the possibility to use either explicit casts or other functions to match the schema of the Sink. This is a completely different issue though.

Reply:
That makes sense to me. I think we can allow legacy decimal type only when the logical precision is (38, 18). What do you think?

Btw, we missed to check logical-physical mapping for sinks in blink planner. We should add that in this PR or a following PR.

Reply:
Thanks @dawidwys for summary.
+1 to add check logical-physical mapping for sink.
+1 to add implicit casts before sink, since that is really a functional regression.

Reply:
Ad. 1 Yes I am checking legacy decimal type against DECIMAL(38,18)

Ad.2 I'd prefer to do that in another PR as this PR would fix the SQL client e2e tests.